pkg/controller/infrastructure/infraflow/ensure_utils.go

```diff
@@ -296,3 +299,145 @@ func isUserRouter(config *gcp.InfrastructureConfig) bool {
 func isUserVPC(config *gcp.InfrastructureConfig) bool {
 	return config.Networks.VPC != nil && len(config.Networks.VPC.Name) > 0
 }
+
+// isSimilar return whether to slices are similar using reflect.DeepEqual
+//
+// Similar in this context means they contain the same elements, not necessarily in the same order.
+// In essence, this is set equality using DeepEqual.
+func isSimilar[T any](s1, s2 []T) bool {
+	if len(s1) != len(s2) {
+		return false
+	}
+	for _, v := range s1 {
+		if !slices.ContainsFunc(s2, func(e T) bool { return reflect.DeepEqual(v, e) }) {
+			return false
+		}
+	}
+	return true
+}
+
+// firewallUpdate returns a firewall rule that reflects the differences from 'left' towards 'right'.
+//
+// The intended use for the returned firewall rule is to serve as a minimal patch when updating
+// firewall rules on gcp.
+func firewallUpdate(left, right *compute.Firewall) (*compute.Firewall, error) {
+	if left == right {
+		// both identical, nothing to do
+		return nil, nil
+	}
+	if left == nil && right != nil {
+		// everything is right, return right
+		return right, nil
+	}
+	if right == nil && left != nil {
+		// strictly speaking, the update here would be an empty Firewall-object with its
+		// ForceSendFields set to all fields. In the intended context of this function,
+		// just return nil signalling that there is nothing to do
+		return nil, nil
+	}
```

Comment:
```
	if left == nil {
		return right, nil
	}
	if right == nil || right == left {
		return nil, nil
	}
```

Although I would consider any case where the `right == nil` either a delete event or an error. I don't think you can even reach this func with a `left == nil` so maybe that's an error too ?

Reply:
These were not reduced for legibility.

Basically these lines attempt to handle the obvious default cases. Thinking about it, I don't think this is the way to go for a function with such a limited scope. I'll replace it with checks for `nil` and require two non-nil inputs.